Swift 4 support

AYastrebov opened the pull request:
- Convert to swift 4
- Remove warnings

Commit:
- convert to swift 4

Commit:
- fixed warnings

hectormatos2011 commented on Sources/Queries/SequenceQuery.swift:
```diff
@@ -49,7 +49,7 @@ public final class SequenceQuery<T: Reflectable>: NilComparable, Matchable {
      - type: The type of the objects found in the collection property being subqueried.
      - subbuilder: A closure that defines queries that describe each object found in the collection property being subqueried. The closure must return an instance of the `SubqueryMatch` enum.
      */
-    @discardableResult public func subquery<U: NSObject>(_ type: U.Type, subbuilder: (_ includeIf: PredicateSubqueryBuilder<U>) -> MatchType) -> FinalizedIncluder<T> where U: Reflectable {
```
Any reason for these removals?

AYastrebov replied:
Tha was done to remove the warning ```SequenceQuery.swift:52:174: Redundant conformance constraint 'U': 'Reflectable'```

hectormatos2011 commented on Sources/Protocols/Reflectable.swift:
```diff
@@ -38,7 +38,7 @@ extension NSObject: Reflectable {
             var propertyNames: [Selector] = []
             for i in 0..<Int(count) {
-                if let propertyName = String(validatingUTF8: property_getName(properties?[i])) {
+                if let propertyName = String(validatingUTF8: property_getName((properties?[i])!)) {
```
I'm typically not a fan of including implicitly unwrapped optionals. Could you pretty please remove it? You could do this as well:

```swift
if let currentProperty = properties?[i], let propertyName = String(validatingUTF8: property_getName(currentProperty))
```

hectormatos2011 commented on Sources/Extensions/NSPredicate+PrediKit.swift:
```diff
@@ -19,7 +19,7 @@ public extension NSPredicate {
      - type: The `Reflectable` class type that you'll be querying against. The type you supply here is what PrediKit will inspect to ensure the property names you specify in your includers are contained in that class' property list.
      - builder: A closure that you use to generate includers that construct each subpredicate in the created `NSPredicate`
      */
-    convenience init<T: Reflectable>(_ type: T.Type, builder: ((_ includeIf: PredicateBuilder<T>) -> Void)) {
```
Why are we removing `Reflectable` here? What was the problem it caused? If I remember correctly, I   put this here to enforce the developer to ensure that whatever type they feed into this conforms to `Reflectable`. The only class conforming to it is `NSObject` at the moment since this library was intended for use with CoreData. Is there a case where you can use PrediKit with straight up Swift-only objects? Is there a particular API where that's needed?

AYastrebov replied:
The same was here ```NSPredicate+PrediKit.swift:22:25: Redundant conformance constraint 'T': 'Reflectable'```

hectormatos2011 replied:
Good to know! Makes sense now too

hectormatos2011 commented on Sources/Queries/MemberQuery.swift:
```diff
@@ -11,7 +11,7 @@ import Foundation
 /**
  A class that facilitates the creation of subqueries against `T`'s custom member properties. Since the creation of this class is initiated in the `PredicateBuilder<T>` class, and this class inherits from it, then member creation is recursive.
  */
-public final class MemberQuery<T: Reflectable, MemberType: Reflectable & AnyObject>: PredicateBuilder<T>, Matchable, NilComparable {
```
See the above question about removing Reflectable. The same case goes for the AnyObject constraints I put here.

AYastrebov replied:
And here ```MemberQuery.swift:14:74: Redundant layout constraint 'MemberType' : 'AnyObject'```

hectormatos2011 reviewed (changes requested):
Thank you so much for putting time into updating this! I would absolutely love to support Swift 4 in PrediKit. So, you're awesome <3

I do, however, have a couple of questions and need a couple of things changed or questions addressed before merging this.

At the moment, it looks like this PR doesn't pass tests. Could you look into which tests are failing and fix them? I generally won't approve a PR if the tests aren't passing.

AYastrebov committed:
- update tvOS to swift 4
- update OS X to swift 4
- update watch OS to swift 4
- update travis yml config file
- update swift-version

AYastrebov committed:
- remove mplicitly unwrapped optional

AYastrebov committed:
- update variables in travis yml

AYastrebov committed:
- add ruby-version file

AYastrebov committed:
- install cocoapods on travis

AYastrebov committed:
- fix apple tv device for travis ci

AYastrebov commented:
@hectormatos2011 tests are passing now. I've updated travis config also

hectormatos2011 commented:
Again, thank you! If I remember someone today I'll make sure to update CocoaPods to reflect the new version
